AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/helper/AndroidManifestFinder.java

```diff
@@ -103,43 +112,102 @@ private File findManifestInSpecifiedPath(String androidManifestPath) throws File
 		return androidManifestFile;
 	}
 
-	/**
-	 * We use a dirty trick to find the AndroidManifest.xml file, since it's not
-	 * available in the classpath. The idea is quite simple : create a fake
-	 * class file, retrieve its URI, and start going up in parent folders to
-	 * find the AndroidManifest.xml file. Any better solution will be
-	 * appreciated.
-	 */
-	private File findManifestInParentsDirectories() throws FileNotFoundException {
-		FileHelper.FileHolder projectRootHolder = FileHelper.findRootProjectHolder(environment.getProcessingEnvironment());
-
-		File projectRoot = projectRootHolder.projectRoot;
-
-		File androidManifestFile = new File(projectRoot, "AndroidManifest.xml");
-		for (int i = 0; i < MAX_PARENTS_FROM_SOURCE_FOLDER; i++) {
-			if (androidManifestFile.exists()) {
-				break;
-			} else {
-				if (projectRoot.getParentFile() != null) {
-					projectRoot = projectRoot.getParentFile();
-					androidManifestFile = new File(projectRoot, "AndroidManifest.xml");
-				} else {
-					break;
-				}
-			}
+	private File findManifestInKnownPaths() throws FileNotFoundException {
+		FileHelper.FileHolder holder = FileHelper.findRootProjectHolder(environment.getProcessingEnvironment());
+		return findManifestInKnownPathsStartingFromGenFolder(holder.sourcesGenerationFolder.getAbsolutePath());
+	}
+
+	File findManifestInKnownPathsStartingFromGenFolder(String sourcesGenerationFolder) throws FileNotFoundException {
+		Matcher gradleMatcher = GRADLE_GEN_FOLDER.matcher(sourcesGenerationFolder);
+		Matcher mavenMatcher = MAVEN_GEN_FOLDER.matcher(sourcesGenerationFolder);
+		Matcher eclipseMatcher = ECLIPSE_GEN_FOLDER.matcher(sourcesGenerationFolder);
+
+		AndroidManifestFinderStrategy strategy = null;
+
+		File androidManifestFile = null;
+
+		if (gradleMatcher.matches()) {
```

Comment:
I think you could add a method `matches(String sourceGenerationFolder` in the implementations of `AndroidManifestFinderStrategy` to avoid the `if..else if..else`. That way, you could create a list a `AndroidManifestFinderStrategy`, iterate it to find the right strategy. The constants `XXX_GEN_FOLDER` would be a inner constant of each strategy.